optuna/distributions.py

```diff
@@ -79,16 +78,31 @@ def _contains(self, param_value_in_internal_repr):
 
         raise NotImplementedError
 
-    @abc.abstractmethod
     def _asdict(self):
         # type: () -> Dict
 
-        raise NotImplementedError
+        return self.__dict__
+
+    def __eq__(self, other):
+        # type: (Any) -> bool
+
+        if not isinstance(other, type(self)):
+            return False
+
+        return self.__dict__ == other.__dict__
+
+    def __ne__(self, other):
+        # type: (Any) -> bool
+
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        # type: () -> int
+
+        return hash(self.__dict__)
```

Comment:
As shown in the following example code, this implementation does not work because Python's `dict` is not hashable. I think it is safe to add unit tests to detect this kind of error.

```python
>>> hash(optuna.distributions.UniformDistribution(0, 1))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "optuna/distributions.py", line 102, in __hash__
    return hash(self.__dict__)
TypeError: unhashable type: 'dict'
```

Reply:
Oops, I fixed the problem and added a unit test (https://github.com/pfnet/optuna/pull/573/commits/c6295ba0105e389bcef5b22e921099d1acb33a3f).